crates/sqlexec/src/planner/physical_plan/create_tunnel_exec.rs

```diff
@@ -0,0 +1,62 @@
+use datafusion::physical_plan::{DisplayAs, ExecutionPlan, Partitioning, Statistics};
+use protogen::metastore::types::options::TunnelOptions;
+
+use crate::planner::EMPTY_SCHEMA;
+
+#[derive(Debug)]
+pub struct CreateTunnelExec {
+    pub name: String,
+    pub if_not_exists: bool,
+    pub options: TunnelOptions,
+}
+
+impl DisplayAs for CreateTunnelExec {
+    fn fmt_as(
+        &self,
+        t: datafusion::physical_plan::DisplayFormatType,
+        f: &mut std::fmt::Formatter,
+    ) -> std::fmt::Result {
+        todo!()
+    }
+}
+
+impl ExecutionPlan for CreateTunnelExec {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn schema(&self) -> datafusion::arrow::datatypes::SchemaRef {
+        EMPTY_SCHEMA.clone()
+    }
+
+    fn output_partitioning(&self) -> datafusion::physical_plan::Partitioning {
+        Partitioning::UnknownPartitioning(1)
+    }
+
+    fn output_ordering(&self) -> Option<&[datafusion::physical_expr::PhysicalSortExpr]> {
+        None
+    }
+
+    fn children(&self) -> Vec<std::sync::Arc<dyn ExecutionPlan>> {
+        vec![]
+    }
+
+    fn with_new_children(
+        self: std::sync::Arc<Self>,
+        children: Vec<std::sync::Arc<dyn ExecutionPlan>>,
+    ) -> datafusion::error::Result<std::sync::Arc<dyn ExecutionPlan>> {
+        Ok(self)
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: std::sync::Arc<datafusion::execution::TaskContext>,
+    ) -> datafusion::error::Result<datafusion::physical_plan::SendableRecordBatchStream> {
+        todo!()
```

Comment:
It seems like in order to use custom physical nodes, we'd have to move a lot of the sessionContext in to native datafusion extensions

```rust
pub struct SessionContext {
    /// The execution client for remote sessions.
    // TODO: This is currently unused, but we'll likely need it for running some
    // of our custom plans on a remote service.
    exec_client: Option<RemoteSessionClient>,
    /// Database catalog.
    catalog: SessionCatalog,
    /// In-memory (temporary) tables.
    current_session_tables: HashMap<String, Arc<MemTable>>,
    /// Native tables.
    tables: NativeTableStorage,
    /// Session variables.
    vars: SessionVars,
    /// Prepared statements.
    prepared: HashMap<String, PreparedStatement>,
    /// Bound portals.
    portals: HashMap<String, Portal>,
    /// Track query metrics for this session.
    metrics: SessionMetrics,
    /// Datafusion session context used for planning and execution.
    df_ctx: DfSessionContext,
    /// Read tables from the environment.
    env_reader: Option<Box<dyn EnvironmentReader>>,
    /// Job runner for background jobs.
    background_jobs: JobRunner,
    /// Specific stuff for remote session.
    remote_ctx: Option<RemoteSessionContext>,
}
```
we'd need to execute the same logic that happens here,
```rust
    pub async fn create_tunnel(&mut self, plan: CreateTunnel) -> Result<()> {
        if let &Some(limit) = self.vars.max_tunnel_count.value() {
            if self.get_tunnel_count() >= limit {
                return Err(ExecError::MaxObjectCount {
                    typ: "tunnels",
                    max: limit,
                    current: self.get_tunnel_count(),
                });
            }
        }

        self.catalog
            .mutate([Mutation::CreateTunnel(service::CreateTunnel {
                name: plan.name,
                if_not_exists: plan.if_not_exists,
                options: plan.options,
            })])
            .await?;
        Ok(())
    }
```
and move it into the `execute` function. 

but that only has access to the datafusion `TaskContext`. So we'd likely have to move the relevent properties into datafusion session extensions.

```rust
fn execute(...) {
  let catalog = context.session_config().get_extension::<SessionCatalog>().unwrap();
  let vars = context.session_config().get_extension::<SessionVars>().unwrap();
}
// ... do rest of logic

Reply:
I actually like the idea of implementing the extensions. We currently have a lot of stuff shoved into `SessionContext` that isn't needed during the actual execution. The extensions would naturally break up stuff needed for planning vs stuff needed for execution (with some amount) of overlap.

And if we go this route, it'd make testing things like `CreateTable` easier since we'd only need a subset for the test.

Reply:
@scsmithr for these extensions, won't we have to throw them into an `Arc<Mutex>` though? 

as of right now, all of our DDL execs just mutate the catalog then return an `EmptyExec`. So wouldn't we need a mutable reference to `catalog` inside the exec?

Reply:
Yeah we would need to wrap in a mutex. Datafusion will automatically wrap in in Arc.

Mutex shouldn't be an issue as they'll be relatively uncontended .